File: katas/typescript/src/OOP/polymorphism/polymorphism-bad.ts
// Violación de Polimorfismo: Lógica condicional para tipos
// ❌ Problema: Usar if/else o switch para diferentes tipos

// ❌ Sin herencia ni polimorfismo - solo objetos planos
interface DogData {
  type: "dog";
  name: string;
}

interface CatData {
  type: "cat";
  name: string;
}

interface BirdData {
  type: "bird";
  name: string;
}

type AnimalData = DogData | CatData | BirdData;

// ❌ Clase que necesita conocer todos los tipos específicos
class AnimalProcessor {
  // ❌ Método lleno de condicionales para cada tipo
  public makeSound(animal: AnimalData): void {
    if (animal.type === "dog") {
      console.log(`${animal.name} dice: ¡Guau guau!`);
    } else if (animal.type === "cat") {
      console.log(`${animal.name} dice: ¡Miau miau!`);
    } else if (animal.type === "bird") {
      console.log(`${animal.name} dice: ¡Pío pío!`);
    }
    // ❌ Si agrego un nuevo animal, debo modificar este método
  }

  // ❌ Más condicionales para cada comportamiento
  public feed(animal: AnimalData): void {
    if (animal.type === "dog") {
      console.log(`${animal.name} está comiendo croquetas`);
    } else if (animal.type === "cat") {
      console.log(`${animal.name} está comiendo pescado`);
    } else if (animal.type === "bird") {
      console.log(`${animal.name} está comiendo semillas`);
    }
  }

  // ❌ Y más condicionales para movimiento
  public move(animal: AnimalData): void {
    if (animal.type === "dog") {
      console.log(`${animal.name} está corriendo`);
    } else if (animal.type === "cat") {
      console.log(`${animal.name} está saltando`);
    } else if (animal.type === "bird") {
      console.log(`${animal.name} está volando`);
    }
  }
}

// ❌ Uso con condicionales por todos lados
console.log("=== Violación de Polimorfismo ===");

const processor = new AnimalProcessor();

const dog: DogData = { type: "dog", name: "Rex" };
const cat: CatData = { type: "cat", name: "Luna" };
const bird: BirdData = { type: "bird", name: "Piolín" };

const animals: AnimalData[] = [dog, cat, bird];

// ❌ El procesador debe verificar el tipo constantemente
animals.forEach((animal) => {
  processor.makeSound(animal);
  processor.feed(animal);
  processor.move(animal);
});

// ❌ Problemas:
// - Muchos if/else y switch repetidos
// - Agregar nuevo animal requiere modificar MUCHOS métodos
// - Propenso a errores (olvidar un caso)
// - Viola Open/Closed Principle
// - Difícil de mantener y escalar

export { AnimalData, DogData, CatData, BirdData, AnimalProcessor };

File: katas/csharp/OOP/polymorphism/polymorphism-bad.cs
using System;

// Violación de Polimorfismo: Lógica condicional para tipos
// ❌ Problema: Usar if/else o switch para diferentes tipos

public class DogData
{
  public string Type = "dog";
  public string Name;
}

public class CatData
{
  public string Type = "cat";
  public string Name;
}

public class BirdData
{
  public string Type = "bird";
  public string Name;
}

// ❌ Clase que necesita conocer todos los tipos específicos
public class AnimalProcessor
{
  // ❌ Método lleno de condicionales para cada tipo
  public void MakeSound(object animal)
  {
    if (animal is DogData dog)
    {
      Console.WriteLine($"{dog.Name} dice: ¡Guau guau!");
    }
    else if (animal is CatData cat)
    {
      Console.WriteLine($"{cat.Name} dice: ¡Miau miau!");
    }
    else if (animal is BirdData bird)
    {
      Console.WriteLine($"{bird.Name} dice: ¡Pío pío!");
    }
    // ❌ Si agrego un nuevo animal, debo modificar este método
  }

  // ❌ Más condicionales para cada comportamiento
  public void Feed(object animal)
  {
    if (animal is DogData dog)
    {
      Console.WriteLine($"{dog.Name} está comiendo croquetas");
    }
    else if (animal is CatData cat)
    {
      Console.WriteLine($"{cat.Name} está comiendo pescado");
    }
    else if (animal is BirdData bird)
    {
      Console.WriteLine($"{bird.Name} está comiendo semillas");
    }
  }

  // ❌ Y más condicionales para movimiento
  public void Move(object animal)
  {
    if (animal is DogData dog)
    {
      Console.WriteLine($"{dog.Name} está corriendo");
    }
    else if (animal is CatData cat)
    {
      Console.WriteLine($"{cat.Name} está saltando");
    }
    else if (animal is BirdData bird)
    {
      Console.WriteLine($"{bird.Name} está volando");
    }
  }
}

// ❌ Uso con condicionales por todos lados
class Program
{
  static void Main()
  {
    Console.WriteLine("=== Violación de Polimorfismo ===");

    var processor = new AnimalProcessor();

    var dog = new DogData { Name = "Rex" };
    var cat = new CatData { Name = "Luna" };
    var bird = new BirdData { Name = "Piolín" };

    var animals = new object[] { dog, cat, bird };

    // ❌ El procesador debe verificar el tipo constantemente
    foreach (var animal in animals)
    {
      processor.MakeSound(animal);
      processor.Feed(animal);
      processor.Move(animal);
    }

    // ❌ Problemas:
    // - Muchos if/else repetidos
    // - Agregar nuevo animal requiere modificar MUCHOS métodos
    // - Propenso a errores (olvidar un caso)
    // - Viola Open/Closed Principle
    // - Difícil de mantener y escalar
  }
}
